sentry-ruby/lib/sentry/faraday.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+require "faraday"
+
+module Sentry
+  module Faraday
+    OP_NAME = "http.client"
+    SPAN_ORIGIN = "auto.http.faraday"
+    BREADCRUMB_CATEGORY = "http"
+
+    module Connection
+      def new(*)
+        connection = super
+
+        unless connection.builder.adapter === ::Faraday::Adapter::NetHttp
+          connection.request :instrumentation, name: OP_NAME, instrumenter: Instrumenter.new
+        end
+
+        # Ensure that the instrumentation middleware is the first one in the stack
+        connection.builder.handlers.reverse!
```

Comment:
We need to do this so that the instrumentation will measure the entire execution stack, not just sending a request, unless we actually do want to only measure the request?

Reply:
okay so this is bad because any monkeypatch has to be _idempotent_, i.e. leave userland behaviour unchanged. This breaks that contract. Instead of `reverse`, can we just make sure our patch is the first in the list?

That doesn't change the ordering of the existing stuff the user might have added.

Reply:
@sl0thentr0py the trick is that we don't know the adapter until we yield. I guess I could add the instrumentation and then remove it if the adapter ends up being net/http. Does it make sense?

Reply:
you can just add the instrumentation always and early return in `instrument` below if the adapter is net/http.
My point is about the `reverse!` because that changes userland behaviour if they also added stuff in the `handlers` list. 
But I also need to read that corresponding faraday code because I don't know how that's working, will get back in a bit.

Reply:
@sl0thentr0py oh d'ooh of course, yeah I just wanted to put it first, but reverse changes the whole handlers array. I will fix it / make it better!